### PR TITLE
SNOW-734920: Cache all types of HTTPS agent

### PR DESCRIPTION
**Pull request by sfc-gh-pmotacki**
### Description
Please explain the changes you made here.

### Checklist
- [ ] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [ ] Create tests which fail without the change (if possible)
- [ ] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [ ] Extend the README / documentation and ensure is properly displayed (if necessary)
- [ ] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

File: lib/global_config.js [hunk 1]
```diff
@@ -255,5 +255,29 @@ function createXmlColumnVariantParser(config) {
...
+
+let keepAlive = true;
+
+/**
+ * Updates the value of the 'keepAlive' parameter.
+ *
+ * @param {boolean} value
+ */
+exports.setKeepAlive = function (value)
+{
```
Comment (sfc-gh-dprzybysz) [inline lib/global_config.js:268]: let's reformat new lines

File: lib/http/base.js [hunk 2]
```diff
@@ -246,19 +224,14 @@ function normalizeHeaders(headers)
     // browser-request will inject its own 'accept': 'application/json' header
     // and the browser XMLHttpRequest object will concatenate the two values and
     // send 'Accept': 'application/json, application/json' with the request
-    var headerNameLowerCase;
-    for (var headerName in headers)
-    {
-      if (headers.hasOwnProperty(headerName))
-      {
+    let headerNameLowerCase;
```
Comment (sfc-gh-dprzybysz) [inline lib/http/base.js:227]: reformatting this file should be in a separate PR - this keep alive feature is important enough to have it without all unimportant changes

File: lib/http/node.js [hunk 3]
```diff
@@ -21,112 +22,96 @@ const Logger = require('../logger');
...
-{
-  const isHttps = Url.parse(url).protocol === 'https:';
-  let agentClass;
-  let agentOptions = { keepAlive: true };
-  let options;
-  var bypassProxy = false;
+const httpsAgentCache = new Map();
+
+function getFromCacheOrCreate(agentClass, options, parsedUrl) {
+  const protocol = parsedUrl.protocol || 'http:';
```
Comment (sfc-gh-dprzybysz) [inline lib/http/node.js:47]: are you sure that `http:` should be the default protocol?

File: test/integration/testLargeResultSet.js [hunk 4]
```diff
@@ -252,7 +252,7 @@ describe('Large result Set Tests', function ()
   });
 });
 
-describe('SNOW-743920: Large result set with ~35 chunks', function () {
+describe('Large result set with ~35 chunks', function () {
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testLargeResultSet.js:255]: I would keep the issue id for the reference

File: test/integration/testManualConnection.js [hunk 5]
```diff
@@ -3,6 +3,7 @@ const async = require("async");
 const assert = require("assert");
 const connOption = require("./connectionOptions");
 const testUtil = require("./testUtil");
+const {performance} = require("perf_hooks");
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testManualConnection.js:6]: can we add this library?

File: test/integration/testLargeResultSet.js [hunk 6]
```diff
@@ -269,7 +269,6 @@ describe('SNOW-743920: Large result set with ~35 chunks', function () {
     await testUtil.executeCmdAsync(connection, 'alter session set USE_CACHED_RESULT = false;');
     await testUtil.executeCmdAsync(connection, createTable);
     await testUtil.executeCmdAsync(connection, populateData);
-    configureLogger('TRACE');
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testLargeResultSet.js:272]: if you remove setting the TRACE level let's remove below setting the ERROR level

File: test/integration/testManualConnection.js [hunk 7]
```diff
@@ -259,4 +260,105 @@ if (process.env.RUN_MANUAL_TESTS_ONLY == "true") {
       });
     });
   });
+
+  describe.only('keepAlive test', function () {
+    this.timeout(180_000);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testManualConnection.js:265]: isn't that default timeout?

File: test/integration/testManualConnection.js [hunk 8]
```diff
@@ -259,4 +260,105 @@ if (process.env.RUN_MANUAL_TESTS_ONLY == "true") {
...
+    });
+    after(async () => {
+      await testUtil.dropTablesIgnoringErrorsAsync(connection, [tableName]);
+      await testUtil.destroyConnectionAsync(connection);
+    });
+
+    it('Verify that requests working faster with keep alive', function (done) {
+      let sumWithKeepAlive = 0;
+      let sumWithoutKeepAlive = 0;
+      async.series(
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testManualConnection.js:287]: could we write this test without async library? with pure async/await?

File: test/integration/testManualConnection.js [hunk 9]
```diff
@@ -259,4 +260,105 @@ if (process.env.RUN_MANUAL_TESTS_ONLY == "true") {
...
+                          resolve();
+                        });
+                    }
+                  }
+                });
+              });
+            }
+          },
+          async function () {
+            snowflake.configure({keepAlive: false});
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testManualConnection.js:322]: I know that it's manual test but it's a global setting and let's turn it off in the after method - maybe in the future we will include it in the CI test suite

File: test/integration/testManualConnection.js [hunk 10]
```diff
@@ -259,4 +260,105 @@ if (process.env.RUN_MANUAL_TESTS_ONLY == "true") {
...
+            connection = snowflake.createConnection(connOption.valid);
+            await testUtil.connectAsync(connection);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testManualConnection.js:324]: both test cases have the same test body - can we create single method to extract common things?

File: lib/global_config.js (hunk 1 above)
Comment (sfc-gh-pmotacki) [reply lib/global_config.js:268]: ok

File: lib/http/base.js (hunk 2 above)
Comment (sfc-gh-pmotacki) [reply lib/http/base.js:227]: ok

File: lib/http/node.js (hunk 3 above)
Comment (sfc-gh-pmotacki) [reply lib/http/node.js:47]: Probably we don't need default here.

File: lib/http/node.js [hunk 11]
```diff
@@ -21,112 +22,96 @@ const Logger = require('../logger');
...
+  if (isHttps) {
+    if (proxy && !bypassProxy) {
+      prepareProxyAgentOptions(agentOptions, proxy);
+      agent = getFromCacheOrCreate(HttpsProxyAgent, agentOptions, parsedUrl);
+    } else {
+      agent = getFromCacheOrCreate(HttpsAgent, agentOptions, parsedUrl);
     }
+  } else if (proxy && !bypassProxy) {
+    prepareProxyAgentOptions(agentOptions, proxy);
+    agent = getFromCacheOrCreate(HttpAgent, agentOptions, url);
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply lib/http/node.js:110]: You're right - parsed url

File: test/integration/ocsp_mock/testConnectionOcspMock.js [hunk 12]
```diff
@@ -9,6 +9,8 @@ const connOption = require('../connectionOptions');
 const Errors = require('../../../lib/errors');
 const ErrorCodes = Errors.codes;
 const HttpsMockAgent = require('./https_ocsp_mock_agent');
+const {configureLogger} = require("../../configureLogger");
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply test/integration/ocsp_mock/testConnectionOcspMock.js:12]: It isn't needed

File: test/integration/testLargeResultSet.js (hunk 6 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testLargeResultSet.js:272]: ok

File: test/integration/testManualConnection.js (hunk 7 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testManualConnection.js:265]: Yes, that's correct

File: test/integration/testManualConnection.js (hunk 9 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testManualConnection.js:322]: ok

File: test/integration/testLargeResultSet.js (hunk 4 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testLargeResultSet.js]: ok

File: test/integration/testManualConnection.js (hunk 5 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testManualConnection.js]: It can be removed

**Review by sfc-gh-dprzybysz — APPROVED** [review]

File: lib/global_config.js [hunk 13]
```diff
@@ -255,5 +255,28 @@ function createXmlColumnVariantParser(config) {
...
+
+let keepAlive = true;
+
+/**
+ * Updates the value of the 'keepAlive' parameter.
+ *
+ * @param {boolean} value
+ */
+exports.setKeepAlive = function (value) {
+  // validate input
```
Comment (sfc-gh-dprzybysz) [inline lib/global_config.js:268]: do we need such comment?

**Review by sfc-gh-dprzybysz — APPROVED** [review]

File: lib/core.js [hunk 14]
```diff
@@ -217,6 +217,15 @@ function Core(options)
         } else if (Util.exists(xmlParserConfig)) {
           GlobalConfig.createXmlColumnVariantParserWithParameters(xmlParserConfig);
         }
+
+        let keepAlive = options.keepAlive;
+        if (Util.exists(keepAlive))
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply lib/core.js:222]: done

File: lib/global_config.js (hunk 13 above)
Comment (sfc-gh-pmotacki) [reply lib/global_config.js:268]: isn't important

File: lib/http/node.js [hunk 15]
```diff
@@ -21,112 +22,107 @@ const Logger = require('../logger');
...
+    httpsAgentCache.set(agentId, agent);
+    Logger.getInstance().trace(`Create and add to cache new agent ${agentId}`);
+    return agent;
+  }
+
+  if (httpsAgentCache.has(agentId)) {
+    Logger.getInstance().trace(`Get agent with id: ${agentId} from cache`);
+    agent = httpsAgentCache.get(agentId);
+    // recreate agent if keepAlive parameter was changed
+    if (agent.keepAlive !== options.keepAlive) {
```
Comment (sfc-gh-dprzybysz) [inline lib/http/node.js:63]: cannot keepAlive be a part of the cache key and then instead of creating new agent we are just reusing agent with turned off keepAlive?

Comment (sfc-gh-pmotacki) [reply lib/http/node.js]: It was first idea ... maybe first is the best ;)
